### PR TITLE
Add implementation for ModeSelectCluster's OnMode attribute and DEPON…

### LINKED ISSUE #11746: Mode Select Cluster's On/Off feature is not supported
#### Problem
 * Mode Select Cluster's On/Off feature should be supported
 * Mode Select Cluster's On/Off feature is not supported

 #### Proposed Solution
Implement the On/Off feature for Mode Select Cluster

### PR DESCRIPTION
**Pull request by du48s03**
…OFF feature.

#### Problem
What is being fixed?  Examples:
* Fixes #11746

#### Change overview
* Add the implementation in src/app/clusters/on-off-server/on-off-server.cpp so that Current-Mode is set to the OnMode when device is changed from Off to On. 

* Add the test for this logic. 

* Add ModeSelectFeature so client can know whether this feature is supported. 
* Reorder the attributes of ModeSelectCluster to match the specification. 
* 
#### Testing
How was this tested? (at least one bullet point required)
* If unit tests were added, how do they cover this issue?
Unit test make sure that the Current Mode is changed to the OnMode value when the device is turned from Off to On.

### CODE REVIEW

File: examples/all-clusters-app/all-clusters-common/all-clusters-app.zap [hunk 1]
```diff
@@ -8023,15 +8023,60 @@
               "mfgCode": null,
               "side": "server",
               "included": 1,
-              "storageOption": "RAM",
+              "storageOption": "NVM",
               "singleton": 0,
               "bounded": 0,
-              "defaultValue": "",
+              "defaultValue": "1",
```
Comment (bzbarsky-apple) [inline examples/all-clusters-app/all-clusters-common/all-clusters-app.zap:8029]: To match current behavior, shouldn't this default value be 0?

File: src/app/clusters/on-off-server/on-off-server.cpp [hunk 2]
```diff
@@ -165,6 +165,22 @@ EmberAfStatus OnOffServer::setOnOffValue(chip::EndpointId endpoint, uint8_t comm
         {
             emberAfOnOffClusterLevelControlEffectCallback(endpoint, newValue);
         }
+#endif
+#ifdef EMBER_AF_PLUGIN_MODE_SELECT
+        // If OnMode is not a null value, then change the current mode to it.
+        uint8_t onMode = 255;
+        status = ModeSelect::Attributes::OnMode::Get(endpoint, &onMode);
```
Comment (bzbarsky-apple) [inline src/app/clusters/on-off-server/on-off-server.cpp:172]: This does not look right.  Isn't OnMode nullable per spec?

File: src/app/clusters/on-off-server/on-off-server.cpp [hunk 3]
```diff
@@ -165,6 +165,22 @@ EmberAfStatus OnOffServer::setOnOffValue(chip::EndpointId endpoint, uint8_t comm
...
+
+        if (status != EMBER_ZCL_STATUS_SUCCESS)
+        {
+            emberAfOnOffClusterPrintln("ERR: reading onMode %x", status);
+            return status;
+        }
+
+        if (onMode != 255) {
```
Comment (bzbarsky-apple) [inline src/app/clusters/on-off-server/on-off-server.cpp:180]: What is this magic 255 number?

File: src/app/tests/suites/TestModeSelectCluster.yaml [hunk 4]
```diff
@@ -88,3 +93,23 @@ tests:
...
+
+    - label: "Toggle OnOff"
+      cluster: "On/Off"
+      command: "on"
+
+    - label: "Verify Current Mode Change"
+      command: "readAttribute"
+      attribute: "CurrentMode"
+      response:
+        value: 7
```
Comment (bzbarsky-apple) [inline src/app/tests/suites/TestModeSelectCluster.yaml:115]: Shouldn't there be a test where we write null to OnMode and check how it behaves?

File: src/app/zap-templates/zcl/data-model/chip/mode-select-cluster.xml [hunk 5]
```diff
@@ -41,11 +41,12 @@ limitations under the License.
...
     <!-- Base data types -->
-    <attribute side="server" code="0x0000" define="CURRENT_MODE"    type="INT8U"                                            writable="false" optional="false" reportable="true">CurrentMode</attribute>
-    <attribute side="server" code="0x0001" define="SUPPORTED_MODES" type="ARRAY" entryType="ModeOptionStruct" length="255"  writable="false" optional="false">SupportedModes</attribute>
-    <attribute side="server" code="0x0002" define="ON_MODE"         type="INT8U"                                            writable="true"  optional="true" default="255">OnMode</attribute>
-    <attribute side="server" code="0x0003" define="START_UP_MODE"   type="INT8U"                                            writable="false" optional="false">StartUpMode</attribute>
-    <attribute side="server" code="0x0004" define="MODE_DESCRIPTION" type="CHAR_STRING"                                     writable="false" optional="false" length="32">Description</attribute>
+    <attribute side="server" code="0x0000" define="MODE_DESCRIPTION" type="CHAR_STRING"                                     writable="false" optional="false" length="32">Description</attribute>
```
Comment (bzbarsky-apple) [inline src/app/zap-templates/zcl/data-model/chip/mode-select-cluster.xml:44]: Where is the 32 coming from?  It's not in the spec....  Does the spec need to be fixed?

File: src/app/zap-templates/zcl/data-model/chip/mode-select-cluster.xml [hunk 6]
```diff
@@ -41,11 +41,12 @@ limitations under the License.
...
+    <attribute side="server" code="0x0001" define="STANDARD_NAMESPACE" type="ENUM16"                                        writable="false" optional="true"  default="65535">StandardNamespace</attribute>
```
Comment (bzbarsky-apple) [inline src/app/zap-templates/zcl/data-model/chip/mode-select-cluster.xml:45]: This is nullable in the spec.  Why is it not nullable here?

File: src/app/zap-templates/zcl/data-model/chip/mode-select-cluster.xml [hunk 7]
```diff
@@ -41,11 +41,12 @@ limitations under the License.
...
+    <attribute side="server" code="0x0002" define="SUPPORTED_MODES" type="ARRAY" entryType="ModeOptionStruct" length="255"  writable="false" optional="false">SupportedModes</attribute>
```
Comment (bzbarsky-apple) [inline src/app/zap-templates/zcl/data-model/chip/mode-select-cluster.xml:46]:
```suggestion
    <attribute side="server" code="0x0002" define="SUPPORTED_MODES" type="ARRAY" entryType="ModeOptionStruct" writable="false" optional="false">SupportedModes</attribute>
```
`length` is meaningless for array attributes.

File: src/app/zap-templates/zcl/data-model/chip/mode-select-cluster.xml [hunk 8]
```diff
@@ -41,11 +41,12 @@ limitations under the License.
...
+    <attribute side="server" code="0x0003" define="CURRENT_MODE"    type="INT8U"                                            writable="false" optional="false" reportable="true">CurrentMode</attribute>
+    <attribute side="server" code="0x0004" define="START_UP_MODE"   type="INT8U"                                            writable="false" optional="false">StartUpMode</attribute>
```
Comment (bzbarsky-apple) [inline src/app/zap-templates/zcl/data-model/chip/mode-select-cluster.xml:48]: This is nullable in the spec.  Why is it not nullable here?

Also, this is optional in the spec.  Why is it not optional here?

File: src/app/zap-templates/zcl/data-model/chip/mode-select-cluster.xml [hunk 9]
```diff
@@ -41,11 +41,12 @@ limitations under the License.
...
+    <attribute side="server" code="0x0005" define="ON_MODE"         type="INT8U"                                            writable="true"  optional="true" default="255">OnMode</attribute>
```
Comment (bzbarsky-apple) [inline src/app/zap-templates/zcl/data-model/chip/mode-select-cluster.xml:49]: This is nullable in the spec.  Why is it not nullable here?

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]

File: src/app/zap-templates/zcl/data-model/chip/mode-select-cluster.xml (hunk 5 above)
Comment (du48s03) [reply src/app/zap-templates/zcl/data-model/chip/mode-select-cluster.xml]: The spec should probably be fixed. It's unbounded at the moment.

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]
This needs to be correctly rebased to tip, without spurious changes to submodules, etc.

**Review by lazarkov — APPROVED** [review]

File: src/app/zap-templates/zcl/data-model/chip/mode-select-cluster.xml (hunk 5 above)
Comment (du48s03) [reply src/app/zap-templates/zcl/data-model/chip/mode-select-cluster.xml]: https://github.com/CHIP-Specifications/connectedhomeip-spec/pull/5028

File: src/app/clusters/on-off-server/on-off-server.cpp [hunk 10]
```diff
@@ -165,6 +166,21 @@ EmberAfStatus OnOffServer::setOnOffValue(chip::EndpointId endpoint, uint8_t comm
...
+#ifdef EMBER_AF_PLUGIN_MODE_SELECT
+        // If OnMode is not a null value, then change the current mode to it.
+        ModeSelect::Attributes::OnMode::TypeInfo::Type onMode;
+        status = ModeSelect::Attributes::OnMode::Get(endpoint, onMode);
+        if (status != EMBER_ZCL_STATUS_SUCCESS)
+        {
+            emberAfOnOffClusterPrintln("ERR: reading onMode %x", status);
+            return status;
+        }
+        if (!onMode.IsNull() && onMode.HasValidValue())
```
Comment (bzbarsky-apple) [inline src/app/clusters/on-off-server/on-off-server.cpp:179]: Why is this checking HasValidValue?  That will never be false here....

**Review by bzbarsky-apple — APPROVED** [review]

File: src/app/clusters/on-off-server/on-off-server.cpp (hunk 10 above)
Comment (du48s03) [reply src/app/clusters/on-off-server/on-off-server.cpp:179]:
```suggestion
        if (!onMode.IsNull())
```

**Review by chrisdecenzo — APPROVED** [review]

**Comment by bzbarsky-apple** [comment]
@du48s03 This PR caused https://github.com/project-chip/connectedhomeip/issues/16954.  Can you please take a look at that?

**Pull request merged**